### PR TITLE
Fiber slice

### LINKED ISSUE #3057: Fiber slice
**Dev issue:** https://github.com/tarantool/tarantool/issues/6085

**Product:** Tarantool
**Since:** 2.11
**Audience/target:** developers
**Root document:** https://www.tarantool.io/en/doc/latest/concepts/coop_multitasking/#cooperative-multitasking,
https://www.tarantool.io/en/doc/latest/reference/reference_lua/fiber/#fiber-module, https://www.tarantool.io/en/doc/latest/reference/reference_lua/compat/fiber_slice_default/
**SME:** @ drewdzzz 

# Details

Introduce fiber slices. Fiber slice is over if fiber is being
executed without yield longer than the value of the slice.
One can check if the slice is over. Also it is checked during
space iteration when one uses select/pairs/get methods and
before DML queries. Fiber slice consists of two parts: warning
and error slices. When warning slice is over, warning
"fiber has not yielded for more than <warn_slice> seconds"
will be written in log, and when error slice is over, an error
"FiberSliceIsExceeded: fiber slice is exceeded" will be
thrown. Default fiber slice is 0.5 sec for warning and 1.0 sec
for error.

Lua API:
slice = {warn = time_in_seconds, err = time_in_seconds}
- fiber.check_slice() - write in warn log if warning slice is over,
throw an error if error slice is over.
- fiber.set_max_slice([fib_id], slice) - set max slice to passed fiber,
otherwise set default max slice that will used by all fibers without
max slice. One can pass number instead of table to set error slice
and disable warning.
- fiber.set_slice(slice) - set slice for current fiber execution.
One can pass number instead of table to set error slice
and disable warning.
- fiber.extend_slice(slice) - extend slice of current fiber execution.
One can pass number instead of table to extend error slice only.

One can send SIGURG to an instance to set current slice to 0.
It will stop iteration over space.

Examples:

Using fiber slice to stop doing stuff.

```lua
fiber = require('fiber')
function fib_f()
    -- Run until condition is false or error slice is over.
    while condition do
        fiber.check_slice()
        do_stuff()
    end
end

fib = fiber.new(fib_f)
-- Set error slice 0.5 sec and no warning slice for our fiber.
fib:set_max_slice(0.5)
fib:set_joinable(true)
fib:join()
```

Using fiber slice to limit iteration in space.

```lua
fiber = require('fiber')
function fib_f()
    -- Full scan on a huge space will be stopped.
    box.space.huge_space:select{}
end

fib = fiber.new(fib_f)
-- Set error timeout 3 sec and warning timeout 0.2 sec for our fiber.
fib:set_max_slice({warn = 0.2, err = 3})
fib:set_joinable(true)
fib:join()
```

## UPDATE

https://github.com/tarantool/tarantool/pull/8338

### Default max slice

The default value for max fiber slice is controlled by the new compat option [`fiber_slice_default`](https://github.com/tarantool/tarantool/wiki/compat:fiber_slice_default) The old default is unlimited (no warnings or errors). The new default is `{warn = 0.5, err = 1.0}`.

### `fiber.info()`

The current effective value of the max fiber slice is reported by `fiber.info()`:

```
tarantool> fiber = require('fiber')
---
...

tarantool> fiber.self():set_max_slice({warn = 0.5, err = 1.0})
---
...

tarantool> fiber.self():info().max_slice
---
- err: 1
  warn: 0.5
...
```

If the max fiber slice is unset, it's not reported.

Requested by @ drewdzzz in https://github.com/tarantool/tarantool/commit/b24791a6cdfa53c30a00d8f596b0c124ad29122d.

# Definition of done

* [ ] Fiber user doc contain info on slices and their use cases
* [ ] Fiber module reference contains slice-related functions and parameters

### PR DESCRIPTION
**Pull request by andreyaksenov**
Reworked the `Fibers` section to some extent and added documentation for a fiber slice:
- Added links to new API members to the [Index](https://docs.d.tarantool.io/en/doc/2.11-fiber-slice/reference/reference_lua/fiber/#index) section.
- Added a new [Limit execution time](https://docs.d.tarantool.io/en/doc/2.11-fiber-slice/reference/reference_lua/fiber/#limit-execution-time) section with links to new API members.
- Mentioned a [fiber slice](https://docs.d.tarantool.io/en/doc/2.11-fiber-slice/concepts/coop_multitasking/#cooperative-multitasking) in the `Cooperative multitasking` section.
- Added API links in the [Default value for max fiber slice](https://docs.d.tarantool.io/en/doc/2.11-fiber-slice/reference/reference_lua/compat/fiber_slice_default/) topic.

### CODE REVIEW

File: doc/concepts/coop_multitasking.rst [hunk 1]
```diff
@@ -74,14 +74,11 @@ Explicit yields
...
+explicit yields: :ref:`fiber.yield() <fiber-yield>` and :ref:`fiber.sleep(t) <fiber-sleep>`.
 
-:ref:`fiber.yield() <fiber-yield>` yields execution to another ``ready`` fiber while putting itself in the ``ready`` state, 
-meaning that it will be executed again as soon as possible while being polite to other fibers 
-waiting for execution.
+* :ref:`fiber.yield() <fiber-yield>` yields execution to another ``ready`` fiber while putting itself in the ``ready`` state, meaning that it will be executed again as soon as possible while being polite to other fibers waiting for execution.
 
-:ref:`fiber.sleep(n) <fiber-sleep>` yields execution to another ``ready`` fiber and puts itself in the ``suspended`` 
-state for time ``t`` until time pass and the event loop wakes up that fiber to the ``ready`` state.
+* :ref:`fiber.sleep(n) <fiber-sleep>` yields execution to another ``ready`` fiber and puts itself in the ``suspended`` state for time ``t`` until time pass and the event loop wakes up that fiber to the ``ready`` state.
```
Comment (drewdzzz) [inline doc/concepts/coop_multitasking.rst]: Off-top: `fiber.sleep(n)`, but argument is renamed to `t` in the description. There is `fiber.sleep(t)` above, so it's better to rename argument to `t` here.

File: doc/reference/reference_lua/compat/fiber_slice_default.rst [hunk 2]
```diff
@@ -30,4 +36,4 @@ or the following error is raised unexpectedly by a ``box`` function
 
 then your application has a fiber that may exceed its slice and fail.
 
-To fix this you need to reset the fiber slice with the ``fiber.set_max_slice()`` function.
+To fix this you need to reset the fiber slice with the :ref:`fiber.set_max_slice() <fiber-set_max_slice>` function.
```
Comment (drewdzzz) [inline doc/reference/reference_lua/compat/fiber_slice_default.rst]: 1. `fiber.extend_slice()` was created for such cases.
2. `fiber.yield()` will solve the problem too. Moreover, the mechanism was introduced to show users that they does not share CPU time with other fibers. So, probably, yield will be a better solution :)

File: doc/reference/reference_lua/fiber.rst [hunk 3]
```diff
@@ -63,6 +63,18 @@ Below is a list of all ``fiber`` functions and members.
     | :ref:`fiber.testcancel()             | Check if the current fiber has  |
     | <fiber-testcancel>`                  | been cancelled                  |
     +--------------------------------------+---------------------------------+
+    | :ref:`fiber.set_max_slice()          | Set the default maximum slice   |
```
Comment (drewdzzz) [inline doc/reference/reference_lua/fiber.rst:66]: There is another version of this method: `fiber:set_max_slice()` - set max slice for a particular fiber.

File: doc/reference/reference_lua/fiber.rst [hunk 4]
```diff
@@ -161,32 +171,119 @@ Below is a list of all ``fiber`` functions and members.
...
+
+        fiber has not yielded for more than 0.500 seconds
+
+*   When an error slice is over, the ``FiberSliceIsExceeded`` error is thrown:
+
+    .. code-block:: console
+
+        FiberSliceIsExceeded: fiber slice is exceeded
+
+    The fiber ends and becomes ``dead``.
```
Comment (drewdzzz) [inline doc/reference/reference_lua/fiber.rst]: That's not how it works. If slice is exceeded, fiber status does NOT change. For example:
```lua
local function fib_f()
    while true do
        while pcall(fiber.check_slice) do
            do_job()
        end
        fiber.yield()
    end
end
```
After every time function finds that fiber time is exceeded, it just does `fiber.yield()`, and the fiber will be continued later. So a fiber based on this function will work infinitely, and will not be `dead` - `check_slice` has no impact on its status.
But if there was no `pcall`, an error would be thrown, and function execution would be stopped. And after throwing an error, fiber would be dead.

File: doc/reference/reference_lua/fiber.rst [hunk 5]
```diff
@@ -161,32 +171,119 @@ Below is a list of all ``fiber`` functions and members.
...
+
+The fiber slice is checked by all functions operating on spaces and indexes, such as :ref:`index_object.select() <box_index-select>`, :ref:`space_object.replace() <box_space-replace>`, and so on.
+You can also use the :ref:`fiber.check_slice() <fiber-check_slice>` function in application code to check whether the slice for the current fiber is over.
+
+The following functions can be used to override the default slice value:
+
+* :ref:`fiber.set_max_slice(slice) <fiber-set_max_slice>` sets the default maximum slice for all fibers.
```
Comment (drewdzzz) [inline doc/reference/reference_lua/fiber.rst:265]: Already wrote above - this method can be used to set default maximum slice for a particular fiber.

File: doc/reference/reference_lua/fiber.rst [hunk 6]
```diff
@@ -63,6 +63,18 @@ Below is a list of all ``fiber`` functions and members.
...
     | <fiber-testcancel>`                  | been cancelled                  |
     +--------------------------------------+---------------------------------+
+    | :ref:`fiber.set_max_slice()          | Set the default maximum slice   |
+    | <fiber-set_max_slice>`               | for all fibers                  |
+    +--------------------------------------+---------------------------------+
+    | :ref:`fiber.set_slice()              | Set a slice for the current     |
+    | <fiber-set_slice>`                   | fiber                           |
+    +--------------------------------------+---------------------------------+
+    | :ref:`fiber.extend_slice()           | Extend a slice for the          |
+    | <fiber-extend_slice>`                | current fiber                   |
```
Comment (drewdzzz) [inline doc/reference/reference_lua/fiber.rst]: `set_slice` and `extend_slice` are not about current fiber. They are about current fiber EXECUTION. Current fiber execution ends after a `fiber.yield()`. Example:
```lua
local function fib_f()
    operation_1()
    fiber.set_slice(5)
    operation_2()
    fiber.yield()
    operation_3()
    fiber.extend_slice(4)
    operation_4()
end
```
Imagine that we have default err slice = 1 second. Then, we will have limitation in 1 second for 1st operation. After that, we set err slice 5 seconds for a current fiber execution, that means that we have limitation in 5 seconds for operation 1 and operation 2.
Then `fiber.yield()` happens, and slices are set to defaults. So, we have 1 sec limitation for operation 3, then slice of current execution is extended to 5 sec (1 sec by default + 4 sec by extent) and we have 5 sec limitation for operation 3 + operation 4.

File: doc/reference/reference_lua/fiber.rst [hunk 7]
```diff
@@ -161,32 +171,119 @@ Below is a list of all ``fiber`` functions and members.
...
+
+The fiber slice is checked by all functions operating on spaces and indexes, such as :ref:`index_object.select() <box_index-select>`, :ref:`space_object.replace() <box_space-replace>`, and so on.
+You can also use the :ref:`fiber.check_slice() <fiber-check_slice>` function in application code to check whether the slice for the current fiber is over.
+
+The following functions can be used to override the default slice value:
+
+* :ref:`fiber.set_max_slice(slice) <fiber-set_max_slice>` sets the default maximum slice for all fibers.
+* :ref:`fiber.set_slice(slice) <fiber-set_slice>` sets the slice for a current fiber.
+* :ref:`fiber.extend_slice(slice) <fiber-extend_slice>` extends the slice for a current fiber.
+
```
Comment (drewdzzz) [inline doc/reference/reference_lua/fiber.rst]: Current fiber execution.

File: doc/reference/reference_lua/fiber.rst [hunk 8]
```diff
@@ -591,6 +779,157 @@ recommended.
...
+        $ tarantool app.lua
+        fiber has not yielded for more than 1.500 seconds
+        FiberSliceIsExceeded: fiber slice is exceeded
+
+
+..  _fiber-set_slice:
+
+..  function:: set_slice(slice)
+
+    Set a slice for the current fiber.
```
Comment (drewdzzz) [inline doc/reference/reference_lua/fiber.rst]: Current fiber execution

File: doc/reference/reference_lua/fiber.rst [hunk 9]
```diff
@@ -591,6 +779,157 @@ recommended.
...
+        $ tarantool app.lua
+        fiber has not yielded for more than 1.500 seconds
+        FiberSliceIsExceeded: fiber slice is exceeded
+
+
+..  _fiber-extend_slice:
+
+..  function:: extend_slice(slice)
+
+    Extend a :ref:`slice <fibers_limit_execution_time>` for the current fiber.
```
Comment (drewdzzz) [inline doc/reference/reference_lua/fiber.rst]: Current fiber execution.

**Review by drewdzzz — CHANGES_REQUESTED** [review]
There are 2 fundamental entities, and we need to clarify the difference between them:
1. `max_slice` - slice, that is saved, and is set when fiber wakes (the first execution, or wake up after `fiber.yield()`). Can be set with `fiber_object:set_max_slice()` - this value will be used as a `max_slice` for `fiber_object`. For all fibers without specified `max_slice`, will be used value, set by `fiber.set_max_slice()`.
2. `slice` - used by other methods. Such a slice is a limit of execution before `fiber.yield()`. After `fiber.yield()`, `slice` will be set to `max_slice` for current fiber. I can increase or change `slice`, but it will not affect the execution after the nearest `fiber.yield()`.

File: doc/reference/reference_lua/fiber.rst (hunk 3 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/fiber.rst:66]: thanks for the info, will add the description with the example

File: doc/reference/reference_lua/fiber.rst (hunk 6 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/fiber.rst]: thanks to the detailed explanation, now I got it

File: doc/reference/reference_lua/compat/fiber_slice_default.rst (hunk 2 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/compat/fiber_slice_default.rst]: Updated in the following way:
```
First, make sure that :ref:`fiber.yield() <fibers_yield_control>` is used for this fiber to transfer control to another fiber.
You can also extend the fiber slice with the :ref:`fiber.extend_slice(slice) <fiber-extend_slice>` function.
```
Does it sound good?

File: doc/concepts/coop_multitasking.rst [hunk 10]
```diff
@@ -74,14 +74,11 @@ Explicit yields
...
+explicit yields: :ref:`fiber.yield() <fiber-yield>` and :ref:`fiber.sleep(t) <fiber-sleep>`.
 
-:ref:`fiber.yield() <fiber-yield>` yields execution to another ``ready`` fiber while putting itself in the ``ready`` state, 
-meaning that it will be executed again as soon as possible while being polite to other fibers 
-waiting for execution.
+* :ref:`fiber.yield() <fiber-yield>` yields execution to another ``ready`` fiber while putting itself in the ``ready`` state, meaning that it will be executed again as soon as possible while being polite to other fibers waiting for execution.
 
-:ref:`fiber.sleep(n) <fiber-sleep>` yields execution to another ``ready`` fiber and puts itself in the ``suspended`` 
-state for time ``t`` until time pass and the event loop wakes up that fiber to the ``ready`` state.
+* :ref:`fiber.sleep(t) <fiber-sleep>` yields execution to another ``ready`` fiber and puts itself in the ``suspended`` state for time ``t`` until time pass and the event loop wakes up that fiber to the ``ready`` state.
```
Comment (xuniq) [inline doc/concepts/coop_multitasking.rst]:
```suggestion
* :ref:`fiber.sleep(t) <fiber-sleep>` yields execution to another ``ready`` fiber and puts itself in the ``suspended`` state for time ``t`` until time passes and the event loop wakes up that fiber to the ``ready`` state.
```

File: doc/reference/reference_lua/compat/fiber_slice_default.rst [hunk 11]
```diff
@@ -3,14 +3,20 @@
...
-It is set with the ``fiber.set_max_slice()`` function.
+The max fiber slice specifies the max fiber execution time without yield before a warning is logged or an error is raised.
+It is set with the :ref:`fiber.set_max_slice() <fiber-set_max_slice>` function.
 The new ``compat`` option – ``fiber_slice_default`` – controls the default value of the max fiber slice.
 
 Old and new behavior
 --------------------
 
 The old default value for the max fiber slice is infinity (no warnings or errors). The new default value is ``{warn = 0.5, err = 1.0}``.
+To use the new behavior, set ``fiber_slice_default`` to ``new`` as follows.
```
Comment (xuniq) [inline doc/reference/reference_lua/compat/fiber_slice_default.rst]:
```suggestion
To use the new behavior, set ``fiber_slice_default`` to ``new`` as follows:
```

File: doc/reference/reference_lua/fiber.rst [hunk 12]
```diff
@@ -209,71 +323,83 @@ recommended.
...
-    :param function-arguments: what will be passed to function
+    :param function-arguments: arguments to be passed to the function
 
     :return: created fiber object
     :rtype: userdata
 
     **Example:**
 
-    ..  code-block:: tarantoolsession
+    The script below shows how to create a fiber using ``fiber.create``:
```
Comment (xuniq) [inline doc/reference/reference_lua/fiber.rst]:
```suggestion
    The script below shows how to create a fiber using ``fiber.new``:
```

File: doc/reference/reference_lua/fiber.rst [hunk 13]
```diff
@@ -591,6 +790,157 @@ recommended.
...
+        function long_operation()
+            while clock.monotonic() - time < 5 do
+                fiber.check_slice()
+                -- Long-running operation ⌛⌛⌛ --
+            end
+        end
+
+        long_operation_fiber = fiber.create(long_operation)
+
+    The output should look as follows.
```
Comment (xuniq) [inline doc/reference/reference_lua/fiber.rst]:
```suggestion
    The output should look as follows:
```

File: doc/reference/reference_lua/fiber.rst [hunk 14]
```diff
@@ -341,21 +511,52 @@ recommended.
...
+                fiber.yield()
+            end
+        end
+
+        for i = 1, 2 do
+            fiber_object = fiber.create(greet)
+            fiber_object:cancel()
+        end
+
+    The output might look as follows.
```
Comment (xuniq) [inline doc/reference/reference_lua/fiber.rst]:
```suggestion
    The output might look as follows:
```

**Review by xuniq — APPROVED** [review]

File: doc/reference/reference_lua/fiber.rst (hunk 12 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/fiber.rst]: thanks for catching that :)

**Pull request merged**